kvbc/src/merkle_tree_db_adapter.cpp

```diff
@@ -211,19 +168,100 @@ BlockId DBKeyManipulator::extractBlockIdFromKey(const Key &key) {
 Hash DBKeyManipulator::extractHashFromLeafKey(const Key &key) {
   constexpr auto keyTypeOffset = sizeof(EDBKeyType) + sizeof(EKeySubtype);
   Assert(key.length() > keyTypeOffset + Hash::SIZE_IN_BYTES);
-  Assert(getDBKeyType(key) == EDBKeyType::Key);
-  Assert(getKeySubtype(key) == EKeySubtype::Leaf);
+  Assert(DBKeyManipulator::getDBKeyType(key) == EDBKeyType::Key);
+  Assert(DBKeyManipulator::getKeySubtype(key) == EKeySubtype::Leaf);
   return Hash{reinterpret_cast<const uint8_t *>(key.data() + keyTypeOffset)};
 }
 
+Version DBKeyManipulator::extractVersionFromStaleKey(const Key &key) {
+  constexpr auto keyTypeOffset = sizeof(EDBKeyType) + sizeof(EKeySubtype);
+  Assert(key.length() >= keyTypeOffset + Version::SIZE_IN_BYTES);
+  Assert(DBKeyManipulator::getDBKeyType(key) == EDBKeyType::Key);
+  Assert(DBKeyManipulator::getKeySubtype(key) == EKeySubtype::Stale);
+  return fromBigEndianBuffer<Version::Type>(key.data() + keyTypeOffset);
+}
+
+Key DBKeyManipulator::extractKeyFromStaleKey(const Key &key) {
+  constexpr auto keyOffset = sizeof(EDBKeyType) + sizeof(EKeySubtype) + Version::SIZE_IN_BYTES;
+  Assert(key.length() > keyOffset);
```

Comment:
It is a good point, I agree it needs additional clarification in code. The thing is that the version-only keys don't exist in the database. They are only used as a placeholder for the search as it might be difficult to have a lower bound on a stale key (they have different types, we might change them in the future and so on). We call `seekAtLeast()` with them and expect that the next key has either the requested version, the next one or is not a stale key at all. This is due to lexicographical ordering and the fact that the version comes first.

Some of that is explained by the comments at:

- https://github.com/vmware/concord-bft/pull/522/files#diff-36ad4028f8b859e9e55f5aa0a0e3794aR96

- https://github.com/vmware/concord-bft/pull/522/files#diff-36ad4028f8b859e9e55f5aa0a0e3794aR589

I will put a comment explaining what exactly these version-only keys are used for.